### DISCUSSION TITLE
tablename ("user") is keyword in postgres; data of created table is inaccessable

### DISCUSSION THREAD

**Discussion by david-shiko** [fastapi-users/fastapi-users#582] [Q&A]
postrgres version: 
PostgreSQL 12.6 (Ubuntu 12.6-0ubuntu0.20.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, 64-bit
fastapi-users version      5.1.3
SQLAlchemy                  1.3.23
starlette                          0.13.6
uvicorn                           0.13.4

Bug:
No data after user registration in `user` table

Reason:
`user` is a keyword in postgres, changing table name is helping

Additional:
By some reason, even with wrong table name (`user`) fastapi-users can trace register users, duplicates, etc. But (at least for me) that data isn't accessable 

Temporary fix:
```
class UserTable(Base, SQLAlchemyBaseUserTable):
    # Bug of fastapi-users, current name of table "user" is keyword in postgres, need to replace for any another
    SQLAlchemyBaseUserTable.__tablename__ = 'users'
    pass
```
Warning! This fix may lead to another problem that I don't know how to solve this error without intervening is the source code
```
sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'oauth_account.user_id' could not find table 'user' with which to generate a foreign key to target column 'id'
``` 

How looks a wrong table despite count of registered users:
```
nasvyazi_api_db=# \dt
                 List of relations
 Schema |     Name      | Type  |       Owner       
--------+---------------+-------+-------------------
 public | notes         | table | nasvyazi_api_user
 public | oauth_account | table | nasvyazi_api_user
 public | user          | table | nasvyazi_api_user
(3 rows)

nasvyazi_api_db=# SELECT * FROM user;
   user   
----------
 postgres
(1 row)

nasvyazi_api_db=#
```

**Comment by frankie567** [comment]
You can read more about this issue here: https://dba.stackexchange.com/questions/75551/returning-rows-in-postgresql-with-a-table-called-user

If you write your query like this (with quotes around the table name), it will work:

```sql
SELECT * FROM "user";
```

SQLAlchemy ensures to properly escape the table name, that's why it works properly inside the code.

Now if you *really* want to change the table name, here is what you can do:

```py
class UserTable(Base, SQLAlchemyBaseUserTable):
    __tablename__ = "users"  # Directly overload the __tablename__ property in your class
```

And for the OAuth foreign key to work:

```py
from fastapi_users.db.sqlalchemy import GUID
from sqlalchemy import ForeignKey
from sqlalchemy.ext.declarative import declared_attr


class OAuthAccount(SQLAlchemyBaseOAuthAccountTable, Base):
    @declared_attr
    def user_id(cls):
        return Column(GUID, ForeignKey("users.id", ondelete="cascade"), nullable=False)  # "users.id" instead of "user.id"
```

**Reply by david-shiko** [reply]
Thank you for help. 
By the way, why did you name the table “user” and not “users”? Which is the default name for such tables.

**Reply by frankie567** [reply]
There isn't clear convention there: some prefer singular, others plural: https://stackoverflow.com/questions/338156/table-naming-dilemma-singular-vs-plural-names

Django, for example, simply use the model name, which is singular. So, I would say: why not?
